AppDB/appscale/datastore/cassandra_env/entity_id_allocator.py

```diff
@@ -0,0 +1,243 @@
+    """
+    if retries < 0:
+      raise AppScaleDBConnectionError('Unable to reserve new block')
+
+    try:
+      last_reserved = self._get_last_reserved()
+    except TRANSIENT_CASSANDRA_ERRORS:
+      raise AppScaleDBConnectionError('Unable to get last reserved ID')
+
+    new_reserved = last_reserved + size
+    if new_reserved > self.max_allowed:
+      raise AppScaleBadArg('Exceeded maximum allocated IDs')
+
+    try:
+      self._set_reserved(last_reserved, new_reserved)
+    except ReservationFailed:
+      return self.allocate_size(size, retries=retries - 1)
+
+    start_id = last_reserved + 1
+    end_id = new_reserved
+    return start_id, end_id
+
+  def allocate_max(self, max_id, retries=5):
+    """ Reserves all IDs up to the one given.
+
+    Args:
+      max_id: An integer specifying the maximum ID to allocated.
+      retries: The number of times to retry the reservation.
+    Returns:
+      A tuple of integers specifying the start and end ID.
+    Raises:
+      AppScaleDBConnectionError if the reservation is tried too many times.
+      AppScaleBadArg if the ID space has been exhausted.
+    """
+    if retries < 0:
+      raise AppScaleDBConnectionError('Unable to reserve new block')
+
+    if max_id > self.max_allowed:
+      raise AppScaleBadArg('Exceeded maximum allocated IDs')
+
+    try:
+      last_reserved = self._get_last_reserved()
+    except TRANSIENT_CASSANDRA_ERRORS:
+      raise AppScaleDBConnectionError('Unable to get last reserved ID')
+
+    # Instead of returning an error, the API returns an invalid range.
+    if last_reserved > max_id:
+      return last_reserved + 1, last_reserved
+
+    try:
+      self._set_reserved(last_reserved, max_id)
+    except ReservationFailed:
+      return self.allocate_max(max_id, retries=retries - 1)
+
+    start_id = last_reserved + 1
+    end_id = max_id
+    return start_id, end_id
+
+
+class ScatteredAllocator(EntityIDAllocator):
+  """ An iterator that generates evenly-distributed entity IDs. """
+  def __init__(self, session, project):
+    """ Creates a new ScatteredAllocator instance. Each project should just
+    have one instance since it reserves a large block of IDs at a time.
+
+    Args:
+      session: A cassandra-driver session.
+      project: A string specifying a project ID.
+    """
+    super(ScatteredAllocator, self).__init__(session, project, scattered=True)
+
+    # The range that this datastore has already reserved for scattered IDs.
+    self.start_id = None
+    self.end_id = None
+
+  def __iter__(self):
+    """ Returns a new iterator object. """
+    return self
+
+  def next(self):
+    """ Generates a new entity ID.
+
+    Returns:
+      An integer specifying an entity ID.
+    """
+    # This function will require a tornado lock when made asynchronous.
+    if self.start_id is None or self.start_id > self.end_id:
+      self.start_id, self.end_id = self.allocate_size(DEFAULT_RESERVATION_SIZE)
```

Comment:
It doesn't allocate allocate a block until the first time it needs a scattered ID, but in practice, yes, each start will likely reserve that many IDs for each active project.

It's definitely a trade-off between allocation frequency and efficiency of ID usage. This size allows around 22 billion starts assuming the servers allocate an ID automatically during their life time.